src/braket/experimental/autoqasm/program/program.py

```diff
@@ -123,6 +125,48 @@ def with_calibrations(self, gate_calibrations: Union[Callable, list[Callable]])
         combined_oqpy_program += self._oqpy_program
         return Program(combined_oqpy_program, has_pulse_control=True)
 
+    def make_bound_program(self, param_values: dict[str, float]) -> Program:
+        """Binds FreeParameters based upon their name and values passed in. If parameters
+        share the same name, all the parameters of that name will be set to the mapped value.
+
+        Args:
+            param_values (dict[str, float]): A mapping of FreeParameter names
+                to a value to assign to them.
+        Returns:
+            Program: Returns a program with all present parameters fixed to their respective
+            values.
+        """
+        # We have to copy the program so that we don't modify the original, unbound program
+        oqpy_program_copy = copy.deepcopy(self._oqpy_program)
+        # Let's break early if we've processed all the parameter assignments
+        params_to_process = set(param_values.keys())
+        for state in oqpy_program_copy.stack:
```

Comment:
It shouldn't be too bad going through these, because all the inputs should be at the start of the program. We break after we've seen the params requested. So, the worst case scenario is if there's a param that doesn't appear in the program.

We can enable a "strict" option like the circuit class has, then check against `program_conversion_context._free_parameters` (would have to pass that at Program creation). That would prevent the slow case.

I also considered trying to maintain a pointer to the IODeclaration when it's created, but that only gives me a reference to the stmt, not the location of the statement in `state.body`, which is what I need here. I could try to use `list.index` for that, but I'm not sure it's worth the extra complex (and hacky) code in `.add_io_declarations`

Reply:
@rmshaffer 
> what if params_to_process still has element(s) at this point? it means the user passed in some variable names that aren't in the program - should we raise an exception, or at least print a warning?

We could add a `strict` option here, like circuits. Shall we? I feel like it's probably what users would prefer.

Reply:
Yes I think having a `strict` option here would make sense.

By the way, the docstring for the `strict` param of `Circuit.make_bound_circuit` seems wrong:

https://github.com/amazon-braket/amazon-braket-sdk-python/blob/4446f4bd03f29d1f7d09e433e2727f13a55b5077/src/braket/circuits/circuit.py#L858-L859

but the behavior is actually that it raises an exception if *any* of the parameters do not appear:

https://github.com/amazon-braket/amazon-braket-sdk-python/blob/4446f4bd03f29d1f7d09e433e2727f13a55b5077/src/braket/circuits/circuit.py#L884-L886

Reply:
Yes, I almost included an update to the circuits docstring here, but I will open a second PR to `main` so we can update that documentation now